File: src/multidimensional-array.cpp
/*
 * This program needs C++11 support.
 */
#include <string>
#include <cmath>
#include <array>
#include <iostream>
#include <cstdlib>


typedef std::array<float, 2> Vector2f;
typedef std::array<std::array<float, 2>, 2> Matrix2x2f;
Vector2f rotate(Vector2f vec, float angle);
std::string toString(Vector2f vec);


int main(int argc, char* argv[]){
    
    Vector2f xUnit = {1.0, 0.0};
    Vector2f unit = rotate(xUnit, M_PI / 6.0);
    
    std::cout << toString(unit) << std::endl;
    
    return EXIT_SUCCESS;
}


Vector2f rotate(Vector2f vec, float angle){
    
    Matrix2x2f rotateMatrix = {{
        { static_cast<float>(cos(angle)), static_cast<float>(- sin(angle)) },
        { static_cast<float>(sin(angle)), static_cast<float>(cos(angle)) }
    }}; // Note: the outer double brackets is necessary. 
    
    Vector2f result = {
        rotateMatrix[0][0] * vec[0] + rotateMatrix[0][1] * vec[1],
        rotateMatrix[1][0] * vec[0] + rotateMatrix[1][1] * vec[1]
    };
    return result;
}

std::string toString(Vector2f vec){
    return "(" + std::to_string(vec[0]) + ", " + std::to_string(vec[1]) + ")";
}

File: src/multidimensional_array.rs
//! multi-dimensional array

/** typings  */
type Vector2f = [f64; 2];
type Matrix2x2f = [Vector2f; 2];


/// rotate a 2-dim vector
fn rotate(vec: &Vector2f, angle: f64) -> Vector2f {

    let rotate_mat: Matrix2x2f = [
        [angle.cos(), -angle.sin()],
        [angle.sin(), angle.cos()]
    ];

    return [
        rotate_mat[0][0] * vec[0] + rotate_mat[0][1] * vec[1],
        rotate_mat[1][0] * vec[0] + rotate_mat[1][1] * vec[1]
    ]
}


/// string representation
fn to_string(vec: &Vector2f) -> String {
    return format!("({:.2}, {:.2})", vec[0], vec[1])
}


/// main method
pub fn main() {
    
    let x_unit: Vector2f = [1.0, 0.0];
    let oblique_unit = rotate(&x_unit, std::f64::consts::PI / 6.0);

    println!("{}", to_string(&oblique_unit));
}
